master/internal/api_project.go

```diff
@@ -37,6 +39,87 @@ func (a *apiServer) GetProjectByID(
 	return p, nil
 }
 
+func (a *apiServer) getProjectColumnsByID(
+	ctx context.Context, id int32, curUser model.User,
+) (*apiv1.GetProjectColumnsResponse, error) {
+	notFoundErr := status.Errorf(codes.NotFound, "project (%d) not found", id)
+	p := &projectv1.Project{}
+	if err := a.m.db.QueryProto("get_project", p, id); errors.Is(err, db.ErrNotFound) {
+		return nil, notFoundErr
+	} else if err != nil {
+		return nil, errors.Wrapf(err, "error fetching project (%d) from database", id)
+	}
+	if ok, err := project.AuthZProvider.Get().CanGetProject(ctx, curUser, p); err != nil {
+		return nil, err
+	} else if !ok {
+		return nil, notFoundErr
+	}
+	// Get general columns
+	generalColumns := make([]projectv1.GeneralColumn, 0, len(projectv1.GeneralColumn_value))
+	for gc := range projectv1.GeneralColumn_value {
+		generalColumns = append(
+			generalColumns, projectv1.GeneralColumn(projectv1.GeneralColumn_value[gc]))
+	}
+	// Get hyperpatameters columns
+	hyperparameters := []struct {
+		Hyperparameters []expconf.Hyperparameters
+	}{}
+	err := db.Bun().
+		NewSelect().Table("exp_hyperparameters").Column("hyperparameters").Where(
+		"project_id = ?", id).Scan(ctx, &hyperparameters)
+	if err != nil {
+		return nil, err
+	}
+	hyperparameterMap := make(map[string]string)
+	for _, hps := range hyperparameters {
+		for _, hp := range hps.Hyperparameters {
+			hpf := expconf.FlattenHPs(hp)
+			for k := range hpf {
+				hyperparameterMap[k] = k
+			}
+		}
+	}
+
+	hyperColumns := make([]string, 0)
+	for _, mn := range hyperparameterMap {
+		hyperColumns = append(hyperColumns, mn)
+	}
+
+	// Get metrics columns
+	metricNames := []struct {
+		Name string
+	}{}
+	metricColumns := make([]string, 0)
+	err = db.Bun().
+		NewSelect().Table("exp_metrics_name").ColumnExpr(
+		"json_array_elements_text(tname) AS name").Distinct().Where(
+		"project_id = ?", id).Scan(ctx, &metricNames)
+	if err != nil {
+		return nil, err
+	}
+	for _, mn := range metricNames {
+		metricColumns = append(metricColumns, fmt.Sprintf("%s.%s", "training", mn.Name))
+	}
+
+	metricNames = []struct {
+		Name string
+	}{}
+	err = db.Bun().
+		NewSelect().Table("exp_metrics_name").ColumnExpr(
+		"json_array_elements_text(vname) AS name").Distinct().Where(
+		"project_id = ?", id).Scan(ctx, &metricNames)
+	if err != nil {
+		return nil, err
+	}
+	for _, mn := range metricNames {
+		metricColumns = append(metricColumns, fmt.Sprintf("%s.%s", "validation", mn.Name))
+	}
```

Comment:
Could we combine these queries, either by doing a union or by discarding duplicate values ourselves after running the query?